packages/salesforcedx-utils-vscode/src/services/telemetry.ts

```diff
@@ -383,4 +385,60 @@ export class TelemetryService {
         .catch(err => console.error(err));
     }
   }
+
+  private getRandomUserId = (): string => randomBytes(20).toString('hex');
+
+  private buildCliTelemetryCommand(): Command {
+    return new SfCommandBuilder()
+      .withArg('telemetry')
+      .withJson()
+      .build();
+  }
+
+  private async executeCliTelemetry(
+    command: Command
+  ): Promise<string> {
+    const workspacepath = workspaceUtils.getRootWorkspacePath();
+    const execution = new CliCommandExecutor(command, { cwd: workspacepath }
+    ).execute();
+    const cmdOutput = new CommandOutput();
+    const result = cmdOutput.getCmdResult(execution);
+    return result;
+  }
+
+  public async getUserId(): Promise<string> {
+    // Defining UserId in globalState and using the same in appInsights reporter.
+    // Assigns cliId to UserId when it's undefined in global state.
+    // cliId is undefined when cli-telemetry variable disable-telemetry is true.
+    let globalStateUserId = this.extensionContext?.globalState.get<
+      string | undefined
+    >('UserId');
+
+    if (globalStateUserId) {
+      return globalStateUserId;
+    }
+
+    globalStateUserId = await this.executeCliTelemetry(
+      this.buildCliTelemetryCommand()
+    )
+      .then((getCliTelemetryData): string => {
+        const cmdResult = JSON.parse(getCliTelemetryData) as {
+          result?: { cliId: string };
+        };
+        return cmdResult?.result?.cliId ?? this.getRandomUserId();
+      })
+      .catch(error => {
+        console.log(
+          `Error: ${error} occurred in retrieving cliId, generating user-id ..`
+        );
+        return this.getRandomUserId();
+      });
+
+    await this.extensionContext?.globalState.update(
```

Comment:
So it Turns out that globalState is sandboxed for each extension.  SO if we set a value in the core extension, it's not going to be set for E4D or any other extension.  I don't think this is a big deal as we're intending to use cliID and if cli telemetry is turned off we will never reach this code flow.  Might be worth adding a comment above this line that is the randonUserId value is used here it will be unique per extension